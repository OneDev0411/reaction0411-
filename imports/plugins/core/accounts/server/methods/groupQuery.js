import { Meteor } from "meteor/meteor";
import { Accounts, Groups } from "/lib/collections";
import { Reaction } from "/lib/api";

/**
 * @name groupQuery
 * @method
 * @summary query the Groups collection and return group data
 * @param {Object} context - an object containing the per-request state
 * @param {String} id - id of group to query
 * @return {Object} group object
 */
export function groupQuery(context, id) {
  const { userId } = context;

  // If the user is an has sufficient permissions, then allow them to find any group by id
  if (Reaction.hasPermission(["owner", "admin", "reaction-accounts"], userId)) {
    // find groups by shop ID
    return Groups.findOne({
      _id: id
    });
  }

  // Otherwise, only let users see groups that they are members of
  const userAccount = Accounts.findOne({
    _id: userId,
    groups: {
      $in: [id]
    }
  });

  if (userAccount) {
    // Query the groups collection to find a group by `id`
    return Groups.findOne({
      _id: id
    });
  }

  // If user is not found, throw an error
  throw new Meteor.Error("access-denied", "User does not have permissions to view group");
}

export function groupsQuery(context, shopId) {
  const { userId } = context;

  if (Reaction.hasPermission(["owner", "admin", "reaction-accounts"], userId)) {
    // find groups by shop ID
    return Promise.resolve(Groups.rawCollection().find({
      shopId
    }));
  }

  const userAccount = Accounts.findOne({
    _id: userId
  }, {
    fields: {
      groups: 1
    }
  });

  if (userAccount) {
    // Query the groups collection to find a group by `id`
    return Promise.resolve(Groups.rawCollection().find({
      _id: userAccount.groups,
      shopId
    }));
  }

  // If user is not found, throw an error
  throw new Meteor.Error("access-denied", "User does not have permissions to view group");
}
