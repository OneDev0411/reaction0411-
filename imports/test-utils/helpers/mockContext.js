const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
