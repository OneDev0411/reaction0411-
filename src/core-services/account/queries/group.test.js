import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import groupQuery from "./group.js";

const fakeShopId = "FAKE_SHOP_ID";
const fakeGroup = { _id: "FAKE_GROUP_ID", name: "fake", shopId: fakeShopId };

beforeEach(() => {
  jest.resetAllMocks();
});

test("throws not-found if the group does not exist", async () => {
  mockContext.collections.Groups.findOne.mockReturnValueOnce(undefined);
  const result = groupQuery(mockContext, fakeGroup._id);
  return expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("returns the group if userHasPermission returns true", async () => {
  mockContext.collections.Groups.findOne.mockReturnValueOnce(fakeGroup);
  mockContext.userHasPermission.mockReturnValueOnce(true);
  const result = await groupQuery(mockContext, fakeGroup._id);
  expect(mockContext.collections.Groups.findOne).toHaveBeenCalledWith({ _id: fakeGroup._id });
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(
    "reaction:legacy:groups",
    "read"
  );
  expect(result).toEqual(fakeGroup);
});

test("returns the group if userHasPermission returns false but the current user is in the group", async () => {
  mockContext.collections.Groups.findOne.mockReturnValueOnce(fakeGroup);
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.account = { groups: [fakeGroup._id] };
  const result = await groupQuery(mockContext, fakeGroup._id);
  delete mockContext.account;
  expect(mockContext.collections.Groups.findOne).toHaveBeenCalledWith({ _id: fakeGroup._id });
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(
    "reaction:legacy:groups",
    "read"
  );
  expect(result).toEqual(fakeGroup);
});

test("throws access-denied if not allowed", async () => {
  mockContext.collections.Groups.findOne.mockReturnValueOnce(fakeGroup);
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);
  const result = groupQuery(mockContext, fakeGroup._id);
  return expect(result).rejects.toThrowErrorMatchingSnapshot();
});
