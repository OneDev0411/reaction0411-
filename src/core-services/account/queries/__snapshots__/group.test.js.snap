// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws access-denied if not allowed 1`] = `"Access Denied"`;

exports[`throws not-found if the group does not exist 1`] = `"There is no group with this ID"`;
